spring-cloud-function-stream/src/main/java/org/springframework/cloud/function/stream/StreamListeningFunctionInvoker.java

```diff
@@ -96,12 +96,22 @@ private String select(Message<?> input) {
 	private Function<Message<?>, Object> convertInput(String name) {
 		Class<?> inputType = functionInspector.getInputType(name);
 		return m -> {
-			if (inputType.isAssignableFrom(m.getPayload().getClass())) {
-				return m.getPayload();
+			if (functionInspector.isMessage(name)) {
+				return MessageBuilder.withPayload(convertPayload(name, inputType, m))
```

Comment:
This is fine for now, but it should be simplified once we are able to register `@StreamListener`-like methods via API (the invoker handler for the listener does this conversion automatically, but can't do it right now because the type is `Message<?>`.